### PR TITLE
Fix last coverities

### PR DESCRIPTION
**Pull request by radare**

### CODE REVIEW

File: libr/core/cbin.c [hunk 1]
```diff
@@ -1029,7 +1029,7 @@ static int srclineCmp(const void *a, const void *b) {
 	if (fa && fb) {
 		return !strcmp (fa, fb);
 	}
-	return (fa && fb && !strcmp (fa, fb));
+	return (a && !b)? 1: -1;
```
[the remark this answers is not in the document]
Comment (radare) [reply libr/core/cbin.c]:
```
$ cat a.c
int main () {
	printf ("%d %d %d\n",
	strcmp ("a", "b"),
	strcmp ("a", "a"),
	strcmp ("b", "a"));
}
$ ./a.out
-1 0 1
$
```

before it was returning 0, which is wrong, because "a" != NULL, and returning 0 means they are equal. now it returns 1 or -1.. and maybe we can say equal if a == NULL && a == b

[the remark this answers is not in the document]
Comment (ret2libc) [reply libr/core/cbin.c]: We could just do ‘return a ==b’. It returns true if they are both null, false otherwise.

File: libr/core/cbin.c [hunk 2]
```diff
@@ -1024,12 +1024,18 @@ static int bin_pdb(RCore *core, int mode) {
 }
 
 static int srclineCmp(const void *a, const void *b) {
+	return r_str_cmp (a, b, -1);
+#if 0
```
Comment (XVilka) [inline libr/core/cbin.c]: Why not just kill it?

**Comment by radare** [comment]
Wip, easier for me to test dont merge yet

File: libr/core/cbin.c (hunk 2 above)
Comment (ret2libc) [reply libr/core/cbin.c]: yes let's remove this please.
`r_str_cmp` seems exactly what we need!

File: libr/util/str.c [hunk 3]
```diff
@@ -753,22 +753,21 @@ R_API int r_str_ccmp(const char *dst, const char *src, int ch) {
 // Compare two strings for the first len bytes. Returns true if they are equal.
 // NOTE: this is not useful as a comparitor, as it returns true or false.
 R_API int r_str_cmp(const char *a, const char *b, int len) {
-	if (!a || !b) {
-		return false;
-	}
-	if (a == b) {
-		return true;
-	}
+	if (!a && !b) return 0;
```
Comment (ret2libc) [inline libr/util/str.c]: why did you change this function to return 0/1/-1? As the comment says, this just returns true/false (yes, -1 should be false but still, why the change?)

File: libr/util/str.c [hunk 4]
```diff
@@ -753,22 +753,21 @@ R_API int r_str_ccmp(const char *dst, const char *src, int ch) {
...
+	if (!a && b) return -1;
+	if (a && !b) return 1;
+	// a && b
+	if (a == b) return 0;
 	if (len < 0) {
 		return strcmp (a, b);
 	}
 	for (;len--;) {
```
Comment (ret2libc) [inline libr/util/str.c]: I think all this loop could be replaced by a `strncmp`

**Comment by radare** [comment]
The old function was crashing when a or b is null. Now in case one of them is null returns 1 or -1 like strcmp do

**Comment by radare** [comment]
Dont trust the comments. EVER. Thy always lie. Strcmp never returned bool, and -1 is need for sorting

**Comment by radare** [comment]
i changed r_list_uniq to use int instead of bool, otherweise we were using the RListComparator in 2 different ways which is wrong because bool != int and -1,0,1 is not true/false :P just to make clear that instead of abusing the same function delegate signature for both behaviours.

the comment is removed because is missleading and comments are bad. i always say that and in this case the comment caused all the pain to the rest of the codebase. dunno who wrote it :?

**Comment by XVilka** [comment]
What about creating a new type for such kind of functions, which will be effectively a typedef to int, but to show the intentions in the function prototype - "cmp" or "compare".

**Comment by radare** [comment]
avra test is xx :?

**Comment by radare** [comment]
That signature is already in a typedef named RListComparator. And adding another like RListEqualComparator seems overwnginnered to me

**Comment by radare** [comment]
i fixed the avra test but cant find the commit now :-{{

**Comment by XVilka** [comment]
Use git reflog

On Aug 12, 2018 2:25 AM, "radare" <notifications@github.com> wrote:

i fixed the avra test but cant find the commit now :-{{

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/radare/radare2/pull/11007#issuecomment-412293140>, or mute
the thread

**Comment by radare** [comment]
dupe of https://github.com/radare/radare2/pull/11027

**Comment by radare** [comment]
found in another branch and merged
